internal/protoc/compiler.go

```diff
@@ -0,0 +1,736 @@
+}
+
+// each value in the slice of string slices is a flag passed to protoc
+// examples:
+// []string{"--go_out=plugins=grpc:."}
+// []string{"--grpc-cpp_out=.", "--plugin=protoc-gen-grpc-cpp=/path/to/foo"}
+func (c *compiler) getPluginFlagSets(protoSet *file.ProtoSet, dirPath string) ([][]string, error) {
+	// if not generating, or there are no plugins, nothing to do
+	if !c.doGen || len(protoSet.Config.Gen.Plugins) == 0 {
+		return nil, nil
+	}
+	pluginFlagSets := make([][]string, 0, len(protoSet.Config.Gen.Plugins))
+	for _, genPlugin := range protoSet.Config.Gen.Plugins {
+		pluginFlagSet, err := getPluginFlagSet(protoSet, dirPath, genPlugin)
+		if err != nil {
+			return nil, err
+		}
+		pluginFlagSets = append(pluginFlagSets, pluginFlagSet)
+	}
+	return pluginFlagSets, nil
+}
+
+func getPluginFlagSet(protoSet *file.ProtoSet, dirPath string, genPlugin settings.GenPlugin) ([]string, error) {
+	protoFlags, err := getPluginFlagSetProtoFlags(protoSet, dirPath, genPlugin)
+	if err != nil {
+		return nil, err
+	}
+	flagSet := []string{fmt.Sprintf("--%s_out=%s", genPlugin.Name, genPlugin.OutputPath.AbsPath)}
+	if len(protoFlags) > 0 {
+		flagSet = []string{fmt.Sprintf("--%s_out=%s:%s", genPlugin.Name, protoFlags, genPlugin.OutputPath.AbsPath)}
+	}
+	if genPlugin.Path != "" {
+		flagSet = append(flagSet, fmt.Sprintf("--plugin=protoc-gen-%s=%s", genPlugin.Name, genPlugin.Path))
+	}
+	return flagSet, nil
+}
+
+// the return value corresponds to CodeGeneratorRequest.Parameter
+// https://github.com/golang/protobuf/blob/b4deda0973fb4c70b50d226b1af49f3da59f5265/protoc-gen-go/plugin/plugin.pb.go#L103
+// this function basically just sets the Mfile=package values for go and gogo plugins
+func getPluginFlagSetProtoFlags(protoSet *file.ProtoSet, dirPath string, genPlugin settings.GenPlugin) (string, error) {
+	// the type just denotes what Well-Known Type map to use from the wkt package
+	// if not go or gogo, we don't have any special automatic handling, so just return what we have
+	if !genPlugin.Type.IsGo() && !genPlugin.Type.IsGogo() {
+		return genPlugin.Flags, nil
+	}
+	if genPlugin.Type.IsGo() && genPlugin.Type.IsGogo() {
+		return "", fmt.Errorf("internal error: plugin %s is both a go and gogo plugin", genPlugin.Name)
+	}
+	var goFlags []string
+	if genPlugin.Flags != "" {
+		goFlags = append(goFlags, genPlugin.Flags)
+	}
+	genGoPluginOptions := protoSet.Config.Gen.GoPluginOptions
+	if !genGoPluginOptions.NoDefaultModifiers {
+		modifiers := make(map[string]string)
+		for subDirPath, protoFiles := range protoSet.DirPathToFiles {
+			// you cannot include the files in the same package in the Mfile=package map
+			// or otherwise protoc-gen-go, protoc-gen-gogo, etc freak out and put
+			// these packages in as imports
+			if subDirPath != dirPath {
+				for _, protoFile := range protoFiles {
+					path, err := filepath.Rel(protoSet.Config.DirPath, protoFile.Path)
+					if err != nil {
+						// TODO: best effort, maybe error
```

Comment:
I'd prefer to remove the "best efforts" wherever possible. Intuitively it seems better to error and have the user fix the issue. What are your thoughts?

Reply:
I agree in general, but this should be revisited. You might have people specifying absolute paths, which is what this handles.